Ncspot needs new auth every day

**Describe the bug**
Every day when I open ncspot it asks for "To fully enable Web API features, you need to perform a second OAuth2 authorization". ~~It could be after every boot, I haven't checked that yet.~~ (It's timeout based, not reboot).

I think last time I used this back in December (had to change jobs and was forced on Windows for half a year 😢) it wasn't like that. Is this expected behaviour?

**System (please complete the following information):**
 - OS: Ubuntu 24.04
 - Terminal: kitty
 - Version: 1.3.4
 - Installed from: cargo

**Backtrace/Debug log**
```
[2026-07-24][07:19:25] [ncspot::authentication] [INFO] Using cached credentials
[2026-07-24][07:19:25] [ncspot::spotify] [DEBUG] No HTTP proxy set
[2026-07-24][07:19:25] [librespot_core::session] [DEBUG] new Session
[2026-07-24][07:19:25] [librespot::component] [DEBUG] new ApResolver
[2026-07-24][07:19:25] [librespot_core::http_client] [DEBUG] Requesting https://apresolve.spotify.com/?type=accesspoint&type=dealer&type=spclient
[2026-07-24][07:19:25] [librespot_core::session] [INFO] Connecting to AP "ap-gew4.spotify.com:4070"
[2026-07-24][07:19:25] [librespot_core::connection] [DEBUG] Connection to AP established.
[2026-07-24][07:19:25] [librespot_core::connection] [DEBUG] Authenticating with AP using AUTHENTICATION_STORED_SPOTIFY_CREDENTIALS
[2026-07-24][07:19:25] [librespot_core::session] [INFO] Authenticated as '...' !
[2026-07-24][07:19:25] [librespot_core::session] [DEBUG] drop Session
[2026-07-24][07:19:25] [librespot_core::session] [DEBUG] drop Dispatch
[2026-07-24][07:19:25] [librespot::component] [DEBUG] drop ApResolver
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "background", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "background", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "primary", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "secondary", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "title", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "highlight", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "highlight_bg", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "highlight_inactive_bg", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "playing", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "playing_selected", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "playing_bg", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "error", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "error_bg", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "statusbar_progress", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "statusbar_progress_bg", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "statusbar", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "statusbar_bg", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "cmdline", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "cmdline_bg", falling back to default
[2026-07-24][07:19:25] [ncspot::theme] [WARN] Failed to parse color in "search_match", falling back to default
[2026-07-24][07:19:25] [ncspot::authentication] [INFO] Access token expired, attempting to refresh..
[2026-07-24][07:19:25] [reqwest::connect] [DEBUG] starting new connection: https://accounts.spotify.com/
[2026-07-24][07:19:31] [ncspot::authentication] [ERROR] Failed to refresh token: Failed to exchange code for access token (Server returned error response: invalid_request: refresh_token must be supplied)
[2026-07-24][07:19:31] [librespot_oauth] [INFO] OAuth server listening on 127.0.0.1:33321
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - Close the Search screen with the Escape key
+- Stop discarding the Web API refresh token on every refresh, which caused
+  a new OAuth2 login to be required daily
 
 ## [1.3.4]
 

--- a/src/authentication.rs
+++ b/src/authentication.rs
@@ -1,10 +1,11 @@
 use std::fs;
 use std::net::TcpListener;
+use std::path::Path;
 
 use librespot_core::authentication::Credentials as RespotCredentials;
 use librespot_core::cache::Cache;
 use librespot_oauth::OAuthClientBuilder;
-use log::{error, info};
+use log::{error, info, warn};
 
 use crate::config::{self, Config};
 use crate::spotify::Spotify;
@@ -133,8 +134,11 @@ pub fn get_rspotify_token() -> Result<rspotify::Token, String> {
             return Ok(t);
         }
 
-        // Token is expired, try to refresh if we have a refresh token
-        if let Some(ref refresh_token) = t.refresh_token {
+        // Token is expired, try to refresh if we have a refresh token. Spotify's refresh
+        // responses may omit the refresh token, in which case it must be reused, so an
+        // empty/missing value here means we don't actually have a usable refresh token.
+        let refresh_token = t.refresh_token.as_deref().filter(|s| !s.is_empty());
+        if let Some(refresh_token) = refresh_token {
             info!("Access token expired, attempting to refresh..");
             let client_builder = OAuthClientBuilder::new(
                 NCSPOT_CLIENT_ID,
@@ -144,8 +148,8 @@ pub fn get_rspotify_token() -> Result<rspotify::Token, String> {
             if let Ok(oauth_client) = client_builder.build() {
                 match oauth_client.refresh_token(refresh_token) {
                     Ok(new_token) => {
-                        let mapped = map_token(new_token);
-                        let _ = fs::write(path, serde_json::to_string_pretty(&mapped).unwrap());
+                        let mapped = map_token(new_token, Some(refresh_token));
+                        write_token(&path, &mapped);
                         return Ok(mapped);
                     }
                     Err(e) => {
@@ -157,7 +161,7 @@ pub fn get_rspotify_token() -> Result<rspotify::Token, String> {
     }
 
     let t = create_rspotify_token()?;
-    let _ = fs::write(path, serde_json::to_string_pretty(&t).unwrap());
+    write_token(&path, &t);
     Ok(t)
 }
 
@@ -175,11 +179,47 @@ pub fn create_rspotify_token() -> Result<rspotify::Token, String> {
 
     oauth_client
         .get_access_token()
-        .map(map_token)
+        .map(|token| map_token(token, None))
         .map_err(|e| e.to_string())
 }
 
-fn map_token(token: librespot_oauth::OAuthToken) -> rspotify::Token {
+/// Write `token` to `path`, logging (rather than silently discarding) any failure, and
+/// restricting file permissions since the file holds a long-lived credential.
+fn write_token(path: &Path, token: &rspotify::Token) {
+    let json = match serde_json::to_string_pretty(token) {
+        Ok(json) => json,
+        Err(e) => {
+            error!("Failed to serialize rspotify token: {e}");
+            return;
+        }
+    };
+
+    if let Err(e) = fs::write(path, json) {
+        error!("Failed to write rspotify token cache: {e}");
+        return;
+    }
+
+    #[cfg(unix)]
+    {
+        use std::os::unix::fs::PermissionsExt;
+        if let Err(e) = fs::set_permissions(path, fs::Permissions::from_mode(0o600)) {
+            warn!("Failed to restrict permissions on rspotify token cache: {e}");
+        }
+    }
+}
+
+/// Map an OAuth token obtained from librespot into the `rspotify::Token` type used for Web
+/// API calls.
+///
+/// Spotify's refresh-token responses may omit `refresh_token` entirely, in which case the
+/// previously issued refresh token must be reused (it has not been rotated or invalidated).
+/// `librespot_oauth` maps that omission to an empty string, so `fallback_refresh_token` is
+/// substituted whenever the token returned by librespot is empty, to avoid ever persisting an
+/// unusable refresh token.
+fn map_token(
+    token: librespot_oauth::OAuthToken,
+    fallback_refresh_token: Option<&str>,
+) -> rspotify::Token {
     let duration = if token.expires_at > std::time::Instant::now() {
         token.expires_at.duration_since(std::time::Instant::now())
     } else {
@@ -187,11 +227,50 @@ fn map_token(token: librespot_oauth::OAuthToken) -> rspotify::Token {
     };
     let expires_in = chrono::Duration::from_std(duration).unwrap_or(chrono::Duration::seconds(0));
 
+    let refresh_token = if token.refresh_token.is_empty() {
+        fallback_refresh_token.map(str::to_string)
+    } else {
+        Some(token.refresh_token)
+    };
+
     rspotify::Token {
         access_token: token.access_token,
         expires_in,
         scopes: std::collections::HashSet::new(),
         expires_at: Some(chrono::Utc::now() + expires_in),
-        refresh_token: Some(token.refresh_token),
+        refresh_token,
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    fn oauth_token(refresh_token: &str) -> librespot_oauth::OAuthToken {
+        librespot_oauth::OAuthToken {
+            access_token: "access".to_string(),
+            refresh_token: refresh_token.to_string(),
+            expires_at: std::time::Instant::now() + std::time::Duration::from_secs(3600),
+            token_type: "Bearer".to_string(),
+            scopes: Vec::new(),
+        }
+    }
+
+    #[test]
+    fn map_token_keeps_new_refresh_token_when_present() {
+        let mapped = map_token(oauth_token("new-refresh-token"), Some("old-refresh-token"));
+        assert_eq!(mapped.refresh_token, Some("new-refresh-token".to_string()));
+    }
+
+    #[test]
+    fn map_token_falls_back_to_previous_refresh_token_when_omitted() {
+        let mapped = map_token(oauth_token(""), Some("old-refresh-token"));
+        assert_eq!(mapped.refresh_token, Some("old-refresh-token".to_string()));
+    }
+
+    #[test]
+    fn map_token_yields_no_refresh_token_when_omitted_without_fallback() {
+        let mapped = map_token(oauth_token(""), None);
+        assert_eq!(mapped.refresh_token, None);
     }
 }

